Add resize method to plugin

jtarricone opened the pull request:
This adds a couple resize methods to the plugin object. As mentioned in the original issue, these return `Deferred` objects.

To test:
* clone this branch
* try invoking the resize methods on plugins, checking that the sidebar width changes as expected,  persists when the plugin is closed or asleep, and that other actions can be chained to the deferred object.

I did the following to test this locally -->
* add the following to `main.js` of the sample identifiy plugin starting after line 34:
``` javascript
var self = this;
var d = this.app.resize.setWidth("large");
d.then(function() { $(self.container).css({ color: "white", background: "black" }); });
d.resolve();
```
That should execute an action when you click on a point on the map to identify it, that will make the plugin sidebar wider and change its colors.

Connects #788

mmcfarland commented:
Can you add that snippet in the example to the actual sample plugin?  Then we can point to it when TNC devs want to use the feature.

mmcfarland commented on src/GeositeFramework/js/Plugin.js:
```diff
@@ -47,7 +47,43 @@ require(['use!Geosite',
                         dispatcher: N.app.dispatcher,
-                        suppressHelpOnStartup: _.partial(suppressHelpOnStartup, model)
+                        suppressHelpOnStartup: _.partial(suppressHelpOnStartup, model),
+                        resize: {
```
The rest of the `app` object properties are references or function executions, can you do the same for `resize` so that it doesn't obscure the remaining arguments to `initialize`?

mmcfarland commented on src/GeositeFramework/js/Plugin.js:
```diff
@@ -47,7 +47,43 @@ require(['use!Geosite',
+                                    }
+                                } else if (typeof (val) === "number") {
+                                    d.then(function() {
+                                        $uiContainer.css({ width: val });
```
If the plugin ever sets this, then subsequently settings `small` or `large` will never work since the `element.style` will always override a class.  The safest approach would be to remove the `width` css anytime a class is set.  That in turn might push the `case` for both `large` and `small` do be DRY'd up into a function.

mmcfarland commented on src/GeositeFramework/js/Plugin.js:
```diff
@@ -47,7 +47,43 @@ require(['use!Geosite',
+                        suppressHelpOnStartup: _.partial(suppressHelpOnStartup, model),
+                        resize: {
+                            adjustBy: function(val) {
+                                var currentWidth = parseFloat($uiContainer.css("width"));
```
Not entirely consequential, but I don't think the `width` property can actually be a float, consider using `parseInt` instead.

jtarricone replied:
I think the CSS lengths have type `<number>`, which is real-valued.

jtarricone commented:
I refactored a bit in light of the comments made, and I added an example on the sample plugin, to show the various invocations of the method and chaining additional actions to it.

kdeloach commented on src/GeositeFramework/js/Plugin.js:
```diff
@@ -32,7 +32,40 @@ require(['use!Geosite',
+                resizers = {
+                    adjustBy: function(val) {
+                        var currentWidth = parseFloat($uiContainer.css("width"));
+                        var d = $.Deferred();
```
Consider using the Dojo Deferred class. The docs also have some good example code to demonstrate how deferred objects should be used.

Ref: http://dojotoolkit.org/reference-guide/1.9/dojo/Deferred.html

mmcfarland commented on src/GeositeFramework/sample_plugins/identify_point/main.js:
```diff
@@ -16,6 +16,14 @@ define(["dojo/_base/declare", "framework/PluginBase"],
 
+                // add buttons to bind resize events to
+                $(this.container).append('<p>The buttons below demonstrate how to use a plugins "resize" method</p>');
+                $(this.container).append('<button class="resize-btn" data-ui-key="resize-ctl-identify-incr">+</button>');
```
These appends can all be chained together:
```
$(this.container)
    .append(...)
    .append(...)
    .append(...)
```

kdeloach commented on src/GeositeFramework/js/Plugin.js:
```diff
@@ -32,7 +32,40 @@ require(['use!Geosite',
+                                    });
+                                    $uiContainer.addClass("sidebar-width-" + w);
+                                };
+                            };
```
We can simplify this a bit since we're supporting a very limited number of sizes. Ex:

```
$uiContainer.css({ width: 'auto' })
    .removeClass('sidebar-width-small')
    .removeClass('sidebar-width-medium')
    .removeClass('sidebar-width-large')
    .addClass('sidebar-width-' + w);
```

kdeloach commented on src/GeositeFramework/js/Plugin.js:
```diff
@@ -32,7 +32,40 @@ require(['use!Geosite',
-                logger = new Logger(pluginName);
+                logger = new Logger(pluginName),
+                resizers = {
+                    adjustBy: function(val) {
```
I don't think we need to support this resize operation. Consider removing.

kdeloach commented on src/GeositeFramework/js/Plugin.js:
```diff
@@ -32,7 +32,40 @@ require(['use!Geosite',
+                                };
+                            };
+                        if (typeof (val) === "string") {
+                            d.then(setter(val));
```
Consider whitelisting the text arguments which are acceptable to this function. I can imagine someone making a small typo such as `setWidth('Large')` and becoming confused as to why nothing happens.

kdeloach commented on src/GeositeFramework/js/Plugin.js:
```diff
@@ -32,7 +32,40 @@ require(['use!Geosite',
+                                    $uiContainer.addClass("sidebar-width-" + w);
+                                };
+                            };
+                        if (typeof (val) === "string") {
```
`typeof` is an operator so parens are not necessary.

kdeloach commented on src/GeositeFramework/sample_plugins/identify_point/main.js:
```diff
@@ -31,6 +39,55 @@ define(["dojo/_base/declare", "framework/PluginBase"],
+                        return _.compose(_.map(_.range(0, n), function (i) {
+                            return _.delay(changeColor, i * 100, (i % 2 === 0) ? color : originalColor);
+                        }));
+                    }
```
Consider removing all of this color/blink code which might be confusing to plugin developers. 

One practical example that we could demonstrate is how to update a fixed width element, after a resize has completed. For example, we could add a table with sample data that is fixed at 100px, 200px, or 300px,  depending on whether the container is small, medium, or large.

mmcfarland replied:
I like this idea - I found the blink stuff a little complex.  We also don't need to show every option, but just the basic api usage - we'll have documentation in the wiki that will be more comprehensive.  I agree that it should be reduced and simplified.

kdeloach commented on src/GeositeFramework/js/Plugin.js:
```diff
@@ -32,7 +32,40 @@ require(['use!Geosite',
+                                $uiContainer.css({ width: val });
+                            });
+                        }
+                        return d.resolve();
```
Until we add animated transitions, you could simulate the transition delay by doing something like this:

```
var d = new Deferred();
setTimeout(function() {
    // Set width here
    d.resolve();
}, 1000);
return d.promise;
```

jtarricone commented on src/GeositeFramework/sample_plugins/identify_point/main.js:
```diff
@@ -31,6 +39,55 @@ define(["dojo/_base/declare", "framework/PluginBase"],
+                            break;
+                        case "resize-ctl-identify-small":
+                            $(el).on("click", function() {
+                                self.app.resize.setWidth("small").then(blink("red", 5));
```
`small` and `large`, at least, are already defined.

mmcfarland replied:
Yeah, you're good with those two classes + the numeric.

kdeloach replied:
Cool, didn't realize those styles were already defined.

jtarricone commented:
I made the example in the Identify plugin simpler and touched up the implementation a bit.

mmcfarland commented on src/GeositeFramework/js/Plugin.js:
```diff
@@ -32,7 +32,35 @@ require(['use!Geosite',
+                                });
+                            }
+                            d.resolve();
+                        }, 1000);
```
I understand the comment to add this, but I don't think we need to actually add an arbitrary delay of one second.  Let's cut this down to `100`, so that the class has been applied but not incur a noticeable lag.

kellyi commented:
I'm testing out this branch in the browser and the identify point resizing seems to work well. There is a slight but noticeable delay in the demo version between when the text updates to say "Width is now large" etc and when the plugin actually does resize -- it seems like maybe half a second between clicking and the resize happening? Wasn't sure whether that was intended behavior here. (Possibly seems related to the comment @mmcfarland just made above).

Also, may be worth considering having the change call `resize(immediate)` on the Esri map where `immediate = true` if we want to get rid of the map flicker on resizing as was done here: https://github.com/CoastalResilienceNetwork/GeositeFramework/pull/812 . The Esri map object has an internal/built-in delay on resizing which can be circumvented be calling `resize(true)`; otherwise it will lag behind the plugin resize.

kellyi commented on src/GeositeFramework/sample_plugins/identify_point/main.js:
```diff
@@ -16,6 +16,14 @@ define(["dojo/_base/declare", "framework/PluginBase"],
+                    .append('<button class="resize-btn" data-ui-key="resize-ctl-identify-large">Large</button>')
+                    .append('<p class="width-textbox"></p>');
+
+
```
Extra whitespace here.

jtarricone commented:
I reduced the time delay and added an `esriMap.resize(true)` invocation, which does really reduce the map flicker.
